### DISCUSSION TITLE
Variable secret name

### DISCUSSION THREAD

**Discussion by TiagoGouvea** [community/community#25334] [Actions]
<p>I need to get the secret, but the secret name will vary from branch. Some like:</p>
<pre><code class="lang-auto">branch=${GITHUB_REF#refs/heads/}
secret-name="secrets.${branch^^}"
secret-value=${{ secrets[$secret-name] }}"
</code></pre>
<p>How could it be done?</p>

**Comment by ghost** [comment]
<p>GitHub automatically redacts secrets printed to the log, but you should avoid printing secrets to the log intentionally.<br>
<a href="https://help.github.com/en/actions/configuring-and-managing-workflows/creating-and-storing-encrypted-secrets#accessing-your-secrets" class="onebox" target="_blank">https://help.github.com/en/actions/configuring-and-managing-workflows/creating-and-storing-encrypted-secrets#accessing-your-secrets</a></p>
<p>Unfortunately, there isn’t a way to echo secrets out in the logs, we’re sorry about that.</p>
<p>We really appreciate feedback on how we can make GitHub even better and the best way to report any feature requests is directly to us through our feedback form: <a href="https://support.github.com/contact/feedback">https://support.github.com/contact/feedback</a></p>
<p>Our roadmap is not publicly visible, so we recommend that you keep an eye on the <a href="https://github.com/blog">GitHub Blog</a> for the latest announcements about new features.</p>

**Comment by TiagoGouvea** [comment]
<p>I’m not issuing with the echo.</p>
<p>My point is how to get a secret, based on a variable name.</p>
<h3>Variable secret name</h3>
<p>Like, I want to get a secret like <code>secrets.MASTER</code>, <code>secrets.DEV</code>, <code>secrets.STAGING</code>, that means I will have a variable with <code>branch</code> name from where I will try to get the secret.</p>
<p>I updated the original message to explain it better.</p>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/tiagogouvea">@tiagogouvea</a>,<br>
You want to set the value of a variable as the name of a secret and then access the secret via this variable, right?<br>
I’m afraid it is difficult to achieve this request. Because when using the expression syntax ( <strong>${{ }}</strong> ) to access a property of contexts, it does not support to access the property name via a variable in the expression, we need to explicitly provide the property’s full name directly in the expression.</p>

**Comment by TiagoGouvea** [comment]
<p>Then it can’t be done. :frowning_face:</p>
<p>Thank you <a class="mention" href="/brightran">@brightran</a> for you response.</p>

**Comment by cschleiden** [comment]
<p><a class="mention" href="/tiagogouvea">@tiagogouvea</a> something like this should work:</p>
<pre><code class="lang-yaml">jobs:    
  first:
    runs-on: [ubuntu-latest]
    
    env:
      SECRET_NAME: my_secret_name
    steps:
      - run: echo $SECRET | rev
        env:
          SECRET: ${{ secrets[env.SECRET_NAME] }}
</code></pre>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/cschleiden">@cschleiden</a> , I tried as your suggestion, but get an empty value. It seems does not work.</p>
<p><a class="mention" href="/tiagogouvea">@tiagogouvea</a> , have you tried it, and does it work on your side?</p>

**Comment by cschleiden** [comment]
<p>Could you post your workflow?</p>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/cschleiden">@cschleiden</a> ,</p>
<ol>
<li>The link below is a workflow I set up for test with reference to your suggestion.<br>
<a href="https://github.com/BrightRan/TestOutputs/actions/runs/130674952/workflow" rel="nofollow noopener">https://github.com/BrightRan/TestOutputs/actions/runs/130674952/workflow</a>
</li>
</ol>
<p>From the output logs of this workflow run, you can see I got empty value,</p>
<pre><code class="lang-auto">secrets.MASTER = 
</code></pre>
<p>rather than,</p>
<pre><code class="lang-auto">secrets.MASTER = ***
</code></pre>
<ol start="2">
<li>I also tried using <strong>jq</strong> to access the secrets from the JSON-type secrets context. But it also does not work.<br>
It seems that we can’t pass a variable to be as the <strong>key</strong> in the <strong>jq filter</strong>.</li>
</ol>
<pre><code class="lang-yaml">- name: access secret
  run: |
    branch=${GITHUB_REF##*/}
    secret_name=${branch^^}

    echo "${{ toJson(secrets) }}" &gt; SECRETS_CONTEXT.json

    secret_value=$(jq -r --arg secret_name $secret_name '.$secret_name' SECRETS_CONTEXT.json)
    echo "secret_value = $secret_value"
</code></pre>
<p>Now, I can’t think of any other ways to meet <a class="mention" href="/tiagogouvea">@tiagogouvea</a>’s request.</p>

**Comment by cschleiden** [comment]
<p><a class="mention" href="/brightran">@brightran</a>: env variables are case sensitive, you have a mismatch between <code>SECRET_NAME</code> and <code>secret_name</code>. Changing your first example to this works for me:</p>
<p>When run on <code>master</code>, outputs the value of a secret called <code>MASTER</code> (secrets are case-insensitive):</p>
<pre><code class="lang-auto">steps:          
      - name: get secret name
        run: |
          branch=${GITHUB_REF##*/}
          echo "::set-env name=secret_name::${branch^^}"
      - name: pass secret value
        run: echo "::set-env name=secret_value::${{ secrets[env.secret_name] }}"          
      - name: access secret
        run: |
          echo "secrets.$secret_name = ${{ secrets[env.secret_name] }}"
          echo "secrets.$secret_name = $secret_value"            
</code></pre>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/cschleiden">@cschleiden</a>,</p>
<blockquote>
<p>env variables are case sensitive</p>
</blockquote>
<p>Aha~, yes, I made a mistake.<br>
After I corrected the case of the env name to be consistent, the workflow works as expected now.<br>
Thanks for your help.</p>
<p><a class="mention" href="/tiagogouvea">@tiagogouvea</a>,<br>
Please try the workflow below as <a class="mention" href="/cschleiden">@cschleiden</a>’s suggestion.</p>
<pre><code class="lang-yaml">jobs:
  job1:
    name: test access secret
    runs-on: ubuntu-latest
    steps:
      - name: get secret name
        run: |
          branch=${GITHUB_REF##*/}
          echo "::set-env name=secret_name::${branch^^}"

      - name: pass secret value
        run: echo "::set-env name=secret_value::${{ secrets[env.secret_name] }}"

      - name: access secret
        run: |
          echo "secrets.$secret_name = ${{ secrets[env.secret_name] }}"
          echo "secrets.$secret_name = $secret_value"
</code></pre>

**Comment by cschleiden** [comment, marked as answer]
<p>In addition, the ref name is also in the <code>github</code> context. So something like this</p>
<p><code>- run: echo ${{ secrets[github.ref] }}</code></p>
<p>could also work.</p>

**Comment by TiagoGouvea** [comment]
<p>It’s really simple!</p>
<p>Thanks <a class="mention" href="/cschleiden">@cschleiden</a> <a class="mention" href="/brightran">@brightran</a> for your help! You are amazing!</p>

**Comment by learnprofile** [comment]
@TiagoGouvea @cschleiden @BrightRan 
 I am trying your approach in my workflow but i am getting this error not sure what i am missing. can you help me with any suggestions


- name: Fetch secret names
        id: fetch-secrets
        env:
          GH_TOKEN: ${{ secrets.Token_PAT }}
        run: |
          secrets_names=$(gh api repos/owner/repo/environments/test/secrets --jq '.secrets[].name')
          echo "Fetched secrets: $secrets_names"
          echo "secrets_names=$secrets_names" >> $GITHUB_ENV

      - name: Set secret values as environment variables
        id: set-secrets
        run: |
          secrets_names="${{ env.secrets_names }}"
          for secret in $secrets_names; do
            secret_value="${{ secrets[secret] }}"
            echo "$secret=$secret_value" >> $GITHUB_ENV

      - name: Fetch and Encode Environment Secrets
        id: encode-secrets
        run: |
          secrets_names="${{ env.secrets_names }}"
          jq_command="jq -n"

          for secret in $secrets_names; do
            secret_value=$(eval echo \$$secret)
            jq_command="$jq_command --arg $secret \"$secret_value\""
          done

          jq_command="$jq_command '{"
          for secret in $secrets_names; do
            jq_command="$jq_command $secret: \$$secret,"
          done
          jq_command="${jq_command%,} }' > secrets.json"

          echo "Executing jq command: $jq_command"
          eval $jq_command
          
          The workflow is not valid
      
Unrecognized named-value: 'secret'. Located at position 9 within expression: secrets[secret]

any suggestions?
